### DISCUSSION TITLE
Randomize a list of questions in three rounds

### DISCUSSION THREAD

**Discussion by xliu382** [jspsych/jsPsych#1182] [Q&A] [answered]
Hi,

I need to design a three-round experiment:
In round 1, participants will answer 40 questions randomly chosen from a list of 60 questions;
In round 2, participants will see the rest 20 questions in the list;
In round 3, participants will answer the same 40 questions seen in round 2.
How can I make it that everyone will see a different set of 40 questions and 20 questions? Here's the code I wrote, but the questions are repetitive now in round 2 and different in round 3 (I only included 6 questions for now just to test):

Thank you!!

[question.txt](https://github.com/jspsych/jsPsych/files/5510807/question.txt)
(Sorry I don't know how to show the code directly rather than attaching a file...

**Comment by becky-gilbert** [comment, marked as answer]
Hi @xliu382, sorry for the delay. Your approach of sampling from the `timeline_variables` array using the `sample` parameter makes sense for situations when you only need to do the sampling once, without re-using the same items. But in your case I would instead do the random sampling outside of the trial procedure, using [jsPsych's randomization functions](https://www.jspsych.org/core_library/jspsych-randomization/), so that you can store the randomly-selected items and re-use them later. 

Here's one approach that might work:
```javascript
// array with all 60 questions
var all_questions = [ 
  {data: { i: 1 }, q: 'Q1'},
  {data: { i: 2 }, q: 'Q2'},
  {data: { i: 3 }, q: 'Q3'},
  {data: { i: 4 }, q: 'Q4'},
  {data: { i: 5 }, q: 'Q5'},
  {data: { i: 6 }, q: 'Q6'},
  // ...
  {data: { i: 60 }, q: 'Q60'},
];
// randomize the array 
var all_questions_random = jsPsych.randomization.shuffle(all_questions);
// use the first 40 questions in the randomized array for rounds 1 and 3
// get the first 40 items using the .slice function: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice
var round_1_questions = all_questions_random.slice(0,39);
// use the last 20 questions in the randomized array for round 2
var round_2_questions = all_questions_random.slice(39);

var round1_trial = {
  timeline_variables: round_1_questions 
  // other trial procedure parameters...
};

var round2_trial = {
  timeline_variables: round_2_questions 
  // other trial procedure parameters...
};

var round3_trial = {
  timeline_variables: round_1_questions 
  // other trial procedure parameters...
};  
```

**Reply by xliu382** [reply]
Thank you! I didn't find the .slice function to do that, so I calculated the difference between two arrays instead and it worked:

`var s1 = answers.map((answer,i) => ({answer, question: questions[i]}));`
`var s2 = jsPsych.randomization.sampleWithoutReplacement(s1, 20);`
`var s3 = s1.filter(function(v){ return s2.indexOf(v) == -1 });`

But I think your solution is more clear and efficient than mine!
